src/main/scala/com/evolution/scache/LoadingCache.scala

```diff
@@ -116,110 +119,199 @@ private[scache] object LoadingCache {
+                                        entry
+                                          .release1
+                                          .start
+                                          .as {
+                                            state
+                                              .deferred
+                                              .getOrError
+                                              .map(_.value)
+                                              .asLeft[V]
+                                              .asRight[A]
+                                          }
+
+                                      def tryPutNewValue: F[Either[A, Either[F[V], V]]] =
+                                        0.tailRecM { counter =>
+                                          ref
+                                            .access
+                                            .flatMap { case (entryRefs, set) =>
+                                              entryRefs
+                                                .get(key)
+                                                .fold {
+                                                  // No entry present in the map, so we add a new one
+                                                  Ref[F]
+                                                    .of[EntryState[F, V]](EntryState.Value(entry))
+                                                    .flatMap { entryRef =>
+                                                      set(entryRefs.updated(key, entryRef)).map {
+                                                        case true =>
+                                                          a
+                                                            .asLeft[Either[F[V], V]]
+                                                            .asRight[Int]
+                                                        case false =>
+                                                          (counter + 1)
+                                                            .asLeft[Either[A, Either[F[V], V]]]
+                                                      }
+                                                    }
+                                                } { entryRef =>
+                                                  entryRef
+                                                    .get
+                                                    .flatMap {
+                                                      case state: EntryState.Value[F, V] =>
+                                                        releaseAndReturnValue(state).map(_.asRight[Int])
+
+                                                      case state: EntryState.Loading[F, V] =>
+                                                        releaseAndReturnLoading(state).map(_.asRight[Int])
+
+                                                      case EntryState.Removed =>
```

Comment:
why don't we just update map here?

Reply:
because we know for sure that map changed (will not contain the ref we just did `get` on), so `set` will always fail.